new "type system concepts" section (#1743)
python/typing: docs/spec/glossary.rst

      :term:`type expression`, sometimes with additional :term:`type qualifiers <type qualifier>`.
      See :ref:`"Type and annotation expression" <annotation-expression>` for details.

   assignable
      If a type ``B`` is "assignable to" a type ``A``, a  type checker should
      not error on the assignment ``x: A = b``, where ``b`` is some expression
      whose type is ``B``. Similarly for function calls and returns: ``f(b)``
      where ``def f(x: A): ...`` and ``return b`` inside ``def f(...) -> A:``
      are both valid (not a type error) if and only if ``B`` is assignable to
      ``A``. In this case ``A`` is "assignable from" ``B``. For :term:`fully
      static types <fully static type>`, "assignable to" is equivalent to
      ":term:`subtype` of" and "assignable from" is equivalent to
      ":term:`supertype` of". For :term:`gradual types <gradual type>`, a type

>       ``B`` is assignable to a type ``A`` if there exist :term:`fully static
>       <fully static type>` :term:`materializations <materialize>` ``A'`` and
>       ``B'`` of ``A`` and ``B``, respectively, such that ``B'`` is a
>       :term:`subtype` of ``A'``. See :ref:`type-system-concepts`.

Reviewer: Too many links in the same paragraph can be distracting for the reader. In the CPython docs, the rule we usually use is to not link to the same thing more than once in the same paragraph. The glossary definitions for "fully static" and "subtype" were both linked to earlier on in the paragraph:

```suggestion
      ``B`` is assignable to a type ``A`` if there exist :term:`fully static
      :term:`materializations <materialize>` ``A'`` and ``B'`` of ``A`` and
      ``B``, respectively, such that ``B'`` is a subtype of ``A'``. See
      :ref:`type-system-concepts`.
```

There's some similar issues in some of the other entries you've added here

   consistent
      Two :term:`fully static types <fully static type>` are "consistent with"
      each other if they are the same type. Two gradual types are "consistent
      with" each other if they could :term:`materialize` to the same type. See
      :ref:`type-system-concepts`.

   consistent subtype
      "Consistent subtype" is synonymous with ":term:`assignable` to" (and
      "consistent supertype" is synonymous with ":term:`assignable` from"). See
      :ref:`type-system-concepts`.

   distribution
      The packaged file which is used to publish and distribute
      a release. (:pep:`426`)

   fully static type
      A type is "fully static" if it does not contain any :term:`gradual form`.
      Fully static types represent a set of possible runtime values. Fully
      static types participate in the :term:`subtype` relation. See
      :ref:`type-system-concepts`.

   gradual form
      A gradual form is a :term:`type expression` which makes the type it is

>       part of not a :term:`fully static type`, but rather represent a set of

Reviewer: ```suggestion
      part of not a :term:`fully static type`, but rather a representation of a set of
```

      possible static types. See :ref:`type-system-concepts`. The primary
      gradual form is :ref:`Any`. The ellipsis (`...`) is a gradual form in
      some, but not all, contexts. It is a gradual form when used in a
      :ref:`Callable` type, and when used in ``tuple[Any, ...]`` (but not in

>       other :ref:`tuple <tuples>` types.)

Reviewer 2: ```suggestion
      other :ref:`tuple <tuples>` types).
```

   gradual type

>       Types in the Python type system are "gradual types". A gradual type may

Reviewer: ```suggestion
      Types in the Python type system are "gradual". A gradual type may
```

Reviewer: Looks like you applied this change but then unapplied it in a followup commit — was that deliberate? :)

Author: Nope, not deliberate -- I think I got myself confused and I thought you'd actually suggested the opposite change :) Will reapply this change.

      be a :term:`fully static type`, or it may be :ref:`Any`, or a type that
      contains :ref:`Any` or another :term:`gradual form`. Unlike :term:`fully

>       static types <fully static type>`, gradual types do not necessarily

Reviewer: Hrm -- in the previous sentence, you say that the set of gradual types is a superset of the set of fully static types. But here in this following sentence, you distinguish between the two concepts in a way that implies (to me) that a type cannot be gradual and also fully static.

(I would also phrase this in the singular rather than plural -- "a gradual type does not necessarily represent" rather than "gradual types do not necessarily represent". It's slightly clearer -- otherwise you _could_ misread the sentence as saying that it's something they collectively represent as a group.)

Author: The recommendation to switch to singular is very good, making that edit.

Saying that a gradual type represents a set of static types does not exclude fully static types; they represent a size-1 set of static types :) But I agree the phrasing here is a bit awkward; I'll try for something better, let me know how successful I am!

Reviewer: Much improved now, thanks!

      where ``def f(x: A): ...`` and ``return b`` inside ``def f(...) -> A:``
      are both valid (not a type error) if and only if ``B`` is assignable to
      ``A``. In this case ``A`` is "assignable from" ``B``. For :term:`fully
      static types <fully static type>`, "assignable to" is equivalent to
      ":term:`subtype` of" and "assignable from" is equivalent to
      ":term:`supertype` of". For :term:`gradual types <gradual type>`, a type
      ``B`` is assignable to a type ``A`` if there exist fully static
      :term:`materializations <materialize>` ``A'`` and ``B'`` of ``A`` and
      ``B``, respectively, such that ``B'`` is a subtype of ``A'``. See
      :ref:`type-system-concepts`.

   consistent
      Two :term:`fully static types <fully static type>` are "consistent with"
      each other if they are the same type. Two gradual types are "consistent
      with" each other if they could :term:`materialize` to the same type. See
      :ref:`type-system-concepts`. If two types are consistent, they are both
      :term:`assignable` to and from each other.

   consistent subtype
      "Consistent subtype" is synonymous with ":term:`assignable` to" (and

>       "consistent supertype" is synonymous with ":term:`assignable` from"). See

Reviewer: ```suggestion
      "consistent supertype" is synonymous with "assignable from"). See
```

      :ref:`type-system-concepts`.

   distribution
      The packaged file which is used to publish and distribute
      a release. (:pep:`426`)

   fully static type
      A type is "fully static" if it does not contain any :term:`gradual form`.
      Fully static types represent a set of possible runtime values. Fully
      static types participate in the :term:`subtype` relation. See
      :ref:`type-system-concepts`.

   gradual form
      A gradual form is a :term:`type expression` which makes the type it is
      part of not a :term:`fully static type`, but rather a representation of a
      set of possible static types. See :ref:`type-system-concepts`. The

>       primary gradual form is :ref:`Any`. The ellipsis (`...`) is a gradual

Reviewer: I think we need double backticks here as it's ReST 

```suggestion
      primary gradual form is :ref:`Any`. The ellipsis (``...``) is a gradual
```

      form in some, but not all, contexts. It is a gradual form when used in a
      :ref:`Callable` type, and when used in ``tuple[Any, ...]`` (but not in
      other :ref:`tuple <tuples>` types).

   gradual type
      Types in the Python type system are "gradual". A gradual type may be a
      :term:`fully static type`, or it may be :ref:`Any`, or a type that
      contains ``Any`` or another :term:`gradual form`. A gradual type does not
      necessarily represent a single set of possible runtime values; instead it
      can represent a set of possible static types (a set of possible sets of

>       possible runtime values!). Gradual types do not participate in the

Reviewer: I also enjoy this, but readers struggling to learn about these concepts for the first time might not 😄

```suggestion
      possible runtime values). Gradual types do not participate in the
```

Author: Ok, fine! 😆

      See :ref:`"Type and annotation expression" <annotation-expression>` for details.

   assignable
      If a type ``B`` is "assignable to" a type ``A``, a  type checker should
      not error on the assignment ``x: A = b``, where ``b`` is some expression
      whose type is ``B``. Similarly for function calls and returns: ``f(b)``
      where ``def f(x: A): ...`` and ``return b`` inside ``def f(...) -> A:``
      are both valid (not a type error) if and only if ``B`` is assignable to
      ``A``. In this case ``A`` is "assignable from" ``B``. For :term:`fully
      static types <fully static type>`, "assignable to" is equivalent to
      ":term:`subtype` of" and "assignable from" is equivalent to
      ":term:`supertype` of". For :term:`gradual types <gradual type>`, a type
      ``B`` is assignable to a type ``A`` if there exist fully static
      :term:`materializations <materialize>` ``A'`` and ``B'`` of ``A`` and
      ``B``, respectively, such that ``B'`` is a subtype of ``A'``. See
      :ref:`type-system-concepts`.

   consistent
      Two :term:`fully static types <fully static type>` are "consistent with"

>       each other if they are the same type. Two gradual types are "consistent

Reviewer 2: Or if they are equivalent?

      with" each other if they could :term:`materialize` to the same type. See
      :ref:`type-system-concepts`. If two types are consistent, they are both
      :term:`assignable` to and from each other.

   consistent subtype
      "Consistent subtype" is synonymous with ":term:`assignable` to" (and
      "consistent supertype" is synonymous with "assignable from"). See
      :ref:`type-system-concepts`.

   distribution
      The packaged file which is used to publish and distribute
      a release. (:pep:`426`)

   equivalent
      Two :term:`fully static types <fully static type>` ``A`` and ``B`` are
      equivalent if ``A`` is a :term:`subtype` of ``B`` and ``B`` is a
      :term:`subtype` of ``A``. This implies that ``A`` and ``B`` represent the

>       same set of possible runtime objects.

Reviewer 2: In concepts.rst you also define the equivalence relation for gradual types; that should be added here. It is important if `assert_type()` is defined in terms of equivalence.

      "consistent with" each other if they could :term:`materialize` to the
      same type. See :ref:`type-system-concepts`. If two types are consistent,
      they are both :term:`assignable` to and from each other.

   consistent subtype
      "Consistent subtype" is synonymous with ":term:`assignable` to" (and
      "consistent supertype" is synonymous with "assignable from"). See
      :ref:`type-system-concepts`.

   distribution
      The packaged file which is used to publish and distribute
      a release. (:pep:`426`)

   equivalent
      Two :term:`fully static types <fully static type>` ``A`` and ``B`` are
      equivalent if ``A`` is a :term:`subtype` of ``B`` and ``B`` is a
      :term:`subtype` of ``A``. This implies that ``A`` and ``B`` represent the
      same set of possible runtime objects. Two gradual types ``A`` and ``B``
      are equivalent if all :term:`materializations <materialize>` of ``A`` are
      also materializations of ``B``, and all materializations of ``B`` are
      also materializations of ``A``.

   fully static type
      A type is "fully static" if it does not contain any :term:`gradual form`.

>       Fully static types represent a set of possible runtime values. Fully

Reviewer: ```suggestion
      A fully static type represents a set of possible runtime values. Fully
```

      type. See :ref:`type-system-concepts`.

   inline
      Inline type annotations are annotations that are included in the
      runtime code using :pep:`526` and
      :pep:`3107` syntax (the filename ends in ``.py``).

   materialize
      A :term:`gradual type` can be materialized to a more static type
      (possibly a :term:`fully static type`) by replacing :ref:`Any` with a
      type, or by replacing the `...` in a :ref:`Callable` type with a list of
      types, or by replacing ``tuple[Any, ...]`` with a specific-length tuple
      type. This materialization relation is key to defining
      :term:`assignability <assignable>` for gradual types. See
      :ref:`type-system-concepts`.

   module
      A file containing Python runtime code or stubbed type information.

   narrow
      A :term:`fully static type` ``B`` is narrower than a fully static type
      ``A`` if ``B`` is a :term:`subtype` of ``A`` and ``B`` is not
      :term:`equivalent` to ``A``. This means that ``B`` represents a proper
      subset of the possible objects represented by ``A``. "Type narrowing" is
      when a type checker infers that a name or expression must have a narrower

>       type at some locations in control flow, due to some runtime check of its

Reviewer 3: Nit: assignments can also result in narrowing.

      be imported from the :py:mod:`typing` module or equivalently from ``typing_extensions``,
      but some special forms are placed in other modules.

   structural
      A structural type (see e.g. :ref:`Protocols`, :ref:`TypedDict`) defines a
      set of values not by their ``__class__``, but by their properties (e.g.
      attributes, methods, dictionary key/value types). :ref:`Callable` types
      are also structural; a callable type is a subtype of another callable
      type based on their signatures, not a subclass relationship. In contrast,
      see :term:`nominal` types.

   stub
      A file containing only type information, empty of runtime code
      (the filename ends in ``.pyi``). See :ref:`stub-files`.

   subtype
      A :term:`fully static type` ``B`` is a subtype of a fully static type
      ``A`` if and only if the set of possible runtime values represented by
      ``B`` is a subset of the set of possible runtime values represented by
      ``A``. For :term:`nominal` types (classes), subtyping is defined by
      inheritance. For :term:`structural` types, subtyping is defined by a
      shared set of attributes/methods or keys. Subtype is the inverse of

>       :term:`supertype`. Types that are not fully static are not a subtype or
>       supertype of any other type, but via :term:`materialization
>       <materialize>` they can be :term:`assignable` to another type. See
>       :ref:`type-system-concepts`.

Reviewer: Either

```suggestion
      :term:`supertype`. A type that is not fully static is not a subtype or
      supertype of any other type, but via :term:`materialization
      <materialize>` can be :term:`assignable` to another type. See
      :ref:`type-system-concepts`.
```

or

```suggestion
      :term:`supertype`. Types that are not fully static are not subtypes or
      supertypes of any other types, but via :term:`materialization
      <materialize>` they can be :term:`assignable` to other types. See
      :ref:`type-system-concepts`.
```

So that you're consistent in whether you're using singular or plural
